eras/conway/impl/test/Test/Cardano/Ledger/Conway/ConwayEpochs.hs

```diff
@@ -0,0 +1,423 @@
+            }
+  pure
+    GovActionId
+      { gaidTxId = txId
+      , gaidGovActionIx = GovActionIx 0
+      }
+
+voteForProposal ::
+  KeyHash 'DRepRole StandardCrypto ->
+  TxId StandardCrypto ->
+  ImpTestM Conway (TxId StandardCrypto)
+voteForProposal khDRep prop = do
+  submitBasicConwayTx "Vote as DRep" $ \tx ->
+    tx
+      & bodyTxL . votingProceduresTxBodyL
+        .~ VotingProcedures
+          ( Map.singleton
+              (DRepVoter $ KeyHashObj khDRep)
+              ( Map.singleton
+                  ( GovActionId
+                      prop
+                      (GovActionIx 0)
+                  )
+                  ( VotingProcedure
+                      { vProcVote = VoteYes
+                      , vProcAnchor = SNothing
+                      }
+                  )
+              )
+          )
+
+constitutionShouldBe :: String -> ImpTestM Conway ()
+constitutionShouldBe cUrl = do
+  constitution <-
+    getsNES $
+      nesEsL . esLStateL . lsUTxOStateL . utxosGovStateL . to getConstitution
+  let Constitution {constitutionAnchor = Anchor {anchorUrl}} =
+        fromMaybe
+          (error "No constitution has been set")
+          constitution
+  impIO $ anchorUrl `shouldBe` fromJust (textToUrl $ T.pack cUrl)
+
+lookupGovActionState :: GovActionId StandardCrypto -> ImpTestM Conway (GovActionState Conway)
+lookupGovActionState aId = do
+  actions <- getsNES $ nesEsL . esLStateL . lsUTxOStateL . utxosGovStateL . snapshotsGovStateL . curGovSnapshotsL . to snapshotActions
+  let
+    accum (Just x) _ = Just x
+    accum Nothing gas@GovActionState {gasId}
+      | gasId == aId = Just gas
+      | otherwise = Nothing
+  pure . fromMaybe (error "Could not find action state") $ foldl' accum Nothing actions
+
+calculateAcceptedRatio :: GovActionId StandardCrypto -> ImpTestM Conway Rational
+calculateAcceptedRatio aId = do
+  eNo <- getsNES nesELL
+  stakeDistr <- getsNES $ nesEsL . esLStateL . lsUTxOStateL . utxosStakeDistrL
+  poolDistr <- getsNES nesPdL
+  drepDistr <- getsNES $ nesEsL . epochStateDRepDistrL
+  drepState <- getsNES $ nesEsL . esLStateL . lsCertStateL . certVStateL . vsDRepsL
+  committeeState <- getsNES $ nesEsL . esLStateL . lsCertStateL . certVStateL . vsCommitteeStateL
+  gas <- lookupGovActionState aId
+  let
+    ratEnv =
+      RatifyEnv
+        { reStakePoolDistr = poolDistr
+        , reStakeDistr = credMap stakeDistr
+        , reDRepState = drepState
+        , reDRepDistr = extractDRepDistr drepDistr
+        , reCurrentEpoch = eNo - 1
+        , reCommitteeState = committeeState
+        }
+  pure $
+    dRepAcceptedRatio @Conway
+      ratEnv
+      (gas ^. gasDRepVotesL)
+      (gasAction gas)
+
+logAcceptedRatio :: GovActionId StandardCrypto -> ImpTestM Conway ()
+logAcceptedRatio aId = do
+  ratio <- calculateAcceptedRatio aId
+  logEntry $ "DRep accepted ratio: " <> show ratio
+
+isGovActionAccepted :: GovActionId StandardCrypto -> ImpTestM Conway Bool
+isGovActionAccepted gaId = do
```

Comment:
Could we make this reuse `calculateAcceptedRatio` somehow?

Reply:
`logAcceptedRatio` already will call `calculateAcceptedRatio`, if I understand correctly.

Reply:
Yeah, and `dRepAccepted` also uses it